doc/design/parallel-node-config.md

```diff
@@ -0,0 +1,254 @@
+DrainRequired    = "Drain_Required"
+DrainMcpPausing  = "Draining_MCP_Pausing"
+DrainMcpPaused   = "Draining_MCP_Paused"
+Draining         = "Draining"
+DrainingComplete = "Draining_Complete"
+```
+
+Drain controller will watch for node annotation, `SriovNetworkNodeState` and `SriovNetworkPoolConfig` changes:
+```golang
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="",resources=SriovNetworkPoolConfig,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="",resources=SriovNetworkNodeState,verbs=get;list;watch;update;patch
+```
+
+Once `SriovNetworkNodeState.Status.DrainStatus` will be marked as `Drain_Required` by config daemon controller add will start
+draining procedure.  Reconcile loop will check if current `Draining` nodes count is greater or equal than
+`maxUnavailable` and will mark it as `Draining` to start draining procedure. If current Draining nodes count equals to 
+`maxUnavailable`, no `SriovNetworkNodeState` updates will be applied in the current reconcile loop.
+
+Once draining and configuration will be finished config daemon will update `SriovNetworkNodeState.Status.DrainStatus` to
+`Idle`.
+
+We can add a webhook to block the change of the configuration when the nodes are draining to a lower number. Without
+validating webhook controller will proceed drain procedure for the current `Draining` nodes until the count of them
+will decrease to specified `maxUnavailable` number.
+
+We need this logic in controller to have it centralized and not to have race condition issues during
+`SriovNetworkNodeStates` update.
+
+Config daemon will be responsible for setting `SriovNetworkNodeState.Status.DrainStatus=Drain_Required` and
+`SriovNetworkNodeState.Status.DrainStatus=DrainComplete` only. It will simplify its implementation. 
+
+### API Extensions
+
+#### Extend existing CR SriovNetworkNodeState
+SriovNetworkNodeState already contains current state of the configuration progress in a `status` field, so we don't need
+to store drain status in node annotation. `SriovNetworkNodeStateStatus` will be extended to contain drain progress
+instead of node annotation:
+
+```golang
+// SriovNetworkNodeStateStatus defines the observed state of SriovNetworkNodeState
+type SriovNetworkNodeStateStatus struct {
+	Interfaces    InterfaceExts `json:"interfaces,omitempty"`
+    // +kubebuilder:validation:Enum=Idle,Drain_Disabled,Drain_Required,Draining_MCP_Pausing,Draining_MCP_Paused,Draining,Draining_Complete
+	DrainStatus    string        `json:"drainStatus,omitempty"`
+    // +kubebuilder:validation:Enum=Idle,Succeeded,Failed,InProgress
+	SyncStatus    string        `json:"syncStatus,omitempty"`
+	LastSyncError string        `json:"lastSyncError,omitempty"`
+}
+```
+
+#### Extend existing CR SriovNetworkPoolConfig
+SriovNetworkPoolConfig is used only for OpenShift to provide configuration for
+OVS Hardware Offloading. We can extend it to add configuration for the drain
```

Comment:
can you add some info that `SriovNetworkPoolConfig` will now be relevant for all cluster types (kubernetes and openshift)

Reply:
done